grove/measurements/term_grouping.py

```diff
@@ -0,0 +1,146 @@
+
+            for k, group in enumerate(group_terms):
+                if commutation_check(group, term):
+                    group_inds[k] += [(i, j)]
+                    group_terms[k] += [term]
+                    break
+            else:
+                # for ... else means loop completed without a `break`
+                # Which means this needs to start its own group.
+                group_inds.append([(i, j)])
+                group_terms.append([term])
+
+    return group_inds
+
+
+def commuting_sets_by_zbasis(pauli_sums):
+    """
+    Computes commuting sets based on terms having the same diagonal basis
+
+    Following the technique outlined in the appendix of arXiv:1704.05018.
+
+    :param pauli_sums:
+    :return:
+    """
+    def _max_key_overlap(item, diagonal_sets, max_qubit):
+        """
+        Calculate the max overlap of item with keys of diagoanl_sets
+
+        Returns a different key if we find any collisions
+
+        This mutates the sets
+
+        :param item:
+        :param diagonal_sets:
+        :return:
+        """
+        hash_ptp = tuple([item[n] for n in range(max_qubit)])
+
+        keys = list(diagonal_sets.keys())
+        #  if there are keys check for collisions if not return updated
+        #  diagonal_set dictionary with the key and term added
+        for key in keys:  # for each key check any collisions
+            for idx, pauli_tensor_element in enumerate(key):
+                if ((pauli_tensor_element != 'I' and hash_ptp[idx] != 'I')
+                   and hash_ptp[idx] != pauli_tensor_element):
+                    #  item has collision with this key
+                    #  so must be in a different in a different key or new key
+                    break
+            else:
+                #  we've gotten to the end without finding a difference!
+                #  that means this key works with this item!
+                #  Now we must select the longer of the two keys
+                #  longer is the key or item with fewer identities
+                new_key = []
+                for ii in range(len(hash_ptp)):
+                    if hash_ptp[ii] != 'I':
+                        new_key.append(hash_ptp[ii])
+                    elif key[ii] != 'I':
+                        new_key.append(key[ii])
+                    else:
+                        new_key.append('I')
+
+                if tuple(new_key) in diagonal_sets.keys():
+                    diagonal_sets[tuple(new_key)].append(item)
+                else:
+                    diagonal_sets[tuple(new_key)] = diagonal_sets[key]
+                    diagonal_sets[tuple(new_key)].append(item)
+                    del diagonal_sets[key]
+                return diagonal_sets
+
+        diagonal_sets[hash_ptp] = [item]
+        return diagonal_sets
+
+    max_qubit = 0
+    for term in pauli_sums:
+        if term.id() != "":
+            max_qubit = max(max_qubit, max(term.get_qubits()))
+    max_qubit += 1  # index from 1
```

Comment:
is all of this to determine the number of qubits? does this only work with contiguous integer indexing?

Reply:
The way I wrote this I need the entire PauliString (including identities).  Thus i need to know how many qubits total I have.  I think I could probably do it without knowing the max qubit index since pauli terms return identity because of how `__getitem__` was set up.